# Block Fingerprinting APIs for Profile Grid
# Blocks Gamepad, Bluetooth, USB, and Sensor APIs that can be used for fingerprinting
# --disable-gamepad-api
# --disable-bluetooth-api
# --disable-usb-api

--- a/third_party/blink/renderer/modules/gamepad/navigator_gamepad.cc
+++ b/third_party/blink/renderer/modules/gamepad/navigator_gamepad.cc
@@ -4,6 +4,7 @@

 #include "third_party/blink/renderer/modules/gamepad/navigator_gamepad.h"

+#include "base/command_line.h"
 #include "base/auto_reset.h"
 #include "build/build_config.h"
 #include "device/gamepad/public/cpp/gamepad_features.h"
@@ -30,6 +31,9 @@

 namespace blink {

+// Command-line switch to disable gamepad API
+const char kDisableGamepadAPI[] = "disable-gamepad-api";
+
 namespace {

 void HasGamepadConnectionChanged(const Member<Gamepad> old_gamepad,
@@ -89,6 +93,14 @@ NavigatorGamepad& NavigatorGamepad::From

 HeapVector<Member<Gamepad>> NavigatorGamepad::getGamepads(
     Navigator& navigator) {
+  // Check if gamepad API is disabled
+  const base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
+  if (command_line->HasSwitch(kDisableGamepadAPI)) {
+    // Return empty gamepad list
+    HeapVector<Member<Gamepad>> empty_gamepads(Gamepads::kItemCount);
+    return empty_gamepads;
+  }
+
   return NavigatorGamepad::From(navigator).Gamepads();
 }

@@ -185,6 +197,12 @@ void NavigatorGamepad::DispatchOneShotGa
 }

 const HeapVector<Member<Gamepad>>& NavigatorGamepad::Gamepads() {
+  // Check if gamepad API is disabled
+  const base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
+  if (command_line->HasSwitch(kDisableGamepadAPI)) {
+    return gamepads_->items();  // Will be empty
+  }
+
   if (StartUpdatingIfAttached()) {
     if (GetFrame()) {
       GetFrame()->GetLocalFrameHostRemote().GamepadHapticsPermissionQuery();
--- a/third_party/blink/renderer/modules/bluetooth/bluetooth.cc
+++ b/third_party/blink/renderer/modules/bluetooth/bluetooth.cc
@@ -4,6 +4,7 @@

 #include "third_party/blink/renderer/modules/bluetooth/bluetooth.h"

+#include "base/command_line.h"
 #include <utility>

 #include "build/chromeos_buildflags.h"
@@ -34,6 +35,9 @@

 namespace blink {

+// Command-line switch to disable Bluetooth API
+const char kDisableBluetoothAPI[] = "disable-bluetooth-api";
+
 namespace {

 void CanonicalizeFilter(
@@ -190,6 +194,16 @@ ScriptPromise<BluetoothDevice> Bluetooth
   if (!GetExecutionContext())
     return EmptyPromise();

+  // Check if Bluetooth API is disabled
+  const base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
+  if (command_line->HasSwitch(kDisableBluetoothAPI)) {
+    exception_state.ThrowSecurityError(
+        "Web Bluetooth API is disabled.");
+    return ScriptPromise<BluetoothDevice>::Reject(
+        script_state,
+        exception_state.GetException());
+  }
+
   // If |[navigator.getGamepads()]()| was called in the same origin within the
   // last few seconds the WebBluetooth dialog gets removed without user
--- a/third_party/blink/renderer/modules/webusb/usb.cc
+++ b/third_party/blink/renderer/modules/webusb/usb.cc
@@ -4,6 +4,7 @@

 #include "third_party/blink/renderer/modules/webusb/usb.h"

+#include "base/command_line.h"
 #include "base/types/expected.h"
 #include "build/build_config.h"
 #include "mojo/public/cpp/bindings/pending_remote.h"
@@ -27,6 +28,9 @@

 namespace blink {

+// Command-line switch to disable USB API
+const char kDisableUSBAPI[] = "disable-usb-api";
+
 namespace {

 const char kFeaturePolicyBlocked[] =
@@ -85,6 +89,17 @@ ScriptPromise<IDLSequence<USBDevice>> US

   EnsureServiceConnection();

+  // Check if USB API is disabled
+  const base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
+  if (command_line->HasSwitch(kDisableUSBAPI)) {
+    // Return empty device list
+    return ToResolvedPromise<IDLSequence<USBDevice>>(
+        script_state,
+        HeapVector<Member<USBDevice>>(),
+        exception_state
+    );
+  }
+
   auto* resolver =
       MakeGarbageCollected<ScriptPromiseResolver<IDLSequence<USBDevice>>>(
           script_state, exception_state.GetContext());
@@ -121,6 +136,15 @@ ScriptPromise<USBDevice> USB::requestDev
     return ScriptPromise<USBDevice>::RejectWithDOMException(
         script_state, exception);
   }
+
+  // Check if USB API is disabled
+  const base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
+  if (command_line->HasSwitch(kDisableUSBAPI)) {
+    exception_state.ThrowSecurityError("WebUSB API is disabled.");
+    return ScriptPromise<USBDevice>::Reject(
+        script_state,
+        exception_state.GetException());
+  }

   if (!options->hasFilters() && !options->hasExclusionFilters()) {
     exception_state.ThrowTypeError(
--- a/third_party/blink/renderer/modules/sensor/sensor.cc
+++ b/third_party/blink/renderer/modules/sensor/sensor.cc
@@ -4,6 +4,7 @@

 #include "third_party/blink/renderer/modules/sensor/sensor.h"

+#include "base/command_line.h"
 #include "base/containers/span.h"
 #include "services/device/public/cpp/generic_sensor/sensor_traits.h"
 #include "services/device/public/mojom/sensor.mojom-blink.h"
@@ -27,6 +28,9 @@

 namespace blink {

+// Command-line switch to disable sensor APIs
+const char kDisableSensorAPIs[] = "disable-sensor-apis";
+
 namespace {
 const double kWaitingIntervalThreshold = 0.01;

@@ -154,6 +158,15 @@ void Sensor::start() {
   if (state_ != SensorState::kIdle)
     return;

+  // Check if sensor APIs are disabled
+  const base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
+  if (command_line->HasSwitch(kDisableSensorAPIs)) {
+    HandleError(
+        DOMExceptionCode::kNotSupportedError,
+        "Sensor APIs are disabled.");
+    return;
+  }
+
   state_ = SensorState::kActivating;
   Activate();
 }
